coreneuron/io/lfp.hpp

```diff
@@ -0,0 +1,217 @@
+     * medium
+     */
+    LFPCalculator(const Point3Ds& seg_start,
+                  const Point3Ds& seg_end,
+                  const std::vector<double>& radius,
+                  const std::vector<SegmentIdTy>& segment_ids,
+                  const Point3Ds& electrodes,
+                  double extra_cellular_conductivity)
+        : segment_ids_(segment_ids) {
+        if (seg_start.size() != seg_end.size()) {
+            throw std::invalid_argument("Different number of segment starts and ends.");
+        }
+        if (seg_start.size() != radius.size()) {
+            throw std::invalid_argument("Different number of segments and radii.");
+        }
+        double f(1.0 / (extra_cellular_conductivity * 4.0 * 3.141592653589));
+
+        m.resize(electrodes.size());
+        for (size_t k = 0; k < electrodes.size(); ++k) {
+            auto& ms = m[k];
+            ms.resize(seg_start.size());
+            for (size_t l = 0; l < seg_start.size(); l++) {
+                ms[l] = getFactor(electrodes[k], seg_start[l], seg_end[l], radius[l], f);
+            }
+        }
+    }
+
+    template <typename Vector>
+    void lfp(const Vector& membrane_current) {
+        std::vector<double> res(m.size());
+        for (size_t k = 0; k < m.size(); ++k) {
+            res[k] = 0.0;
+            auto& ms = m[k];
+            for (size_t l = 0; l < ms.size(); l++) {
+                res[k] += ms[l] * membrane_current[segment_ids_[l]];
+            }
+        }
+        std::vector<double> res_reduced(res.size());
+#if NRNMPI
+        int mpi_sum{1};
+        nrnmpi_dbl_allreduce_vec(res.data(), res_reduced.data(), res.size(), mpi_sum);
+#else
+        res_reduced = res;
+#endif
+        lfp_values = res_reduced;
+    }
+
+    std::vector<double> lfp_values;
+
+  private:
+    inline F getFactor(const Point3D& e_pos,
+                       const Point3D& seg_0,
+                       const Point3D& seg_1,
+                       const F radius,
+                       const F f) const;
+
+    std::vector<std::vector<double>> m;
+    const std::vector<SegmentIdTy>& segment_ids_;
+};
+
+template <>
+F LFPCalculator<LineSource>::getFactor(const Point3D& e_pos,
+                                       const Point3D& seg_0,
+                                       const Point3D& seg_1,
+                                       const F radius,
+                                       const F f) const {
+    return line_source_lfp_factor(e_pos, seg_0, seg_1, radius, f);
+}
+
+template <>
+F LFPCalculator<PointSource>::getFactor(const Point3D& e_pos,
+                                        const Point3D& seg_0,
+                                        const Point3D& seg_1,
+                                        const F radius,
+                                        const F f) const {
+    return point_source_lfp_factor(e_pos, barycenter(seg_0, seg_1), radius, f);
+}
+
+};  // namespace coreneuron
+
+#endif  // AREA_LFP_H
```

Comment:
comment doesn't match the macro name in prelude. You should use `#pragma once` anyway.

Reply:
Done.